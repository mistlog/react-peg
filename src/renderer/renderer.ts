import { Chunk, isChunk } from "../chunk/chunk";
import { RuleTranslator } from "../translator/rule/rule";
import * as peg from "pegjs";

class Renderer {
    initializer: string;

    start: Chunk;
    grammar: Map<string, string>;
    rules: Map<string, Function>;
    actions: Map<string, Function>;

    constructor(chunk: Chunk) {
        this.initializer = `
            {
                const { actions } = options;
                const globalFunction = {
                    location,
                    text
                };
            }
        `;

        this.start = chunk;
        this.grammar = new Map<string, string>();
        this.grammar.set("initializer", this.initializer);
        this.rules = new Map<string, Function>();
        this.actions = new Map<string, Function>();
    }

    render() {
        this.collectRule(this.start);
        const grammar = [...this.grammar.values()].join("\n\n");
        return grammar;
    }

    collectRule(chunk: Chunk) {
        if (!isChunk(chunk)) {
            return;
        }

        if (chunk.rule) {
            const translator = new RuleTranslator();
            const ruleStr = translator.translate(chunk, this.actions);
            const { name, chunkOutput, rule } = translator.cache;

            if (this.rules.has(name)) {
                return;
            }

            this.rules.set(name, rule);
            this.grammar.set(name, ruleStr);
            this.collectRule(chunkOutput);
        }

        chunk.children.forEach(child => {
            this.collectRule(child as Chunk);
        })
    }
}

class Parser {
    parser: peg.Parser;
    renderer: Renderer;
    grammar: string;

    constructor(renderer: Renderer) {
        this.renderer = renderer;
        this.grammar = this.renderer.render();
        this.parser = peg.generate(this.grammar);
    }

    parse(code: string) {
        try {
            const ast = this.parser.parse(code, { actions: this.renderer.actions });
            return ast;
        } catch (error) {
            throw { message: error.message, location: error.location, stack: error.stack };
        }
    }
}

export function render(chunk: Chunk) {
    const renderer = new Renderer(chunk);
    const parser = new Parser(renderer);
    return parser;
}

