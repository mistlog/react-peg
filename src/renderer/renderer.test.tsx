import { ReactPeg } from "../index";
import { ITracer } from "./renderer";

export type DigitNode = string;

export function Digit() {
    return (
        <pattern action={({ globalFunction }) => globalFunction.text()}>
            <set>0-9</set>
        </pattern>
    );
}


test("renderer", () => {
    const parser = ReactPeg.render(<Digit />);
    const ast = parser.parse("1");
    expect(ast).toEqual("1");
})


test("renderer.parse", () => {
    const parser = ReactPeg.render(<Digit />);
    expect(() => {
        parser.parse("a")
    }).toThrowErrorMatchingSnapshot();
})


test("renderer.cache-rule", () => {

    function CacheTest() {
        return (
            <pattern action={({ values }) => {
                return values.join("");
            }}>
                <list label="values">
                    <Digit />
                    <Digit />
                </list>
            </pattern>
        );
    }

    const parser = ReactPeg.render(<CacheTest />);
    const ast = parser.parse("12");
    expect(ast).toEqual("12");
})

test("renderer: tracer", () => {
    function TracerTest() {
        return (
            <pattern action={({ values }) => {
                return values.join("");
            }}>
                <list label="values">
                    <Digit />
                    <Digit />
                    <Digit />
                </list>
            </pattern>
        );
    }

    const tracer: ITracer = {
        trace: event => {
            expect(event).toMatchSnapshot();
        }
    };
    const parser = ReactPeg.render(<TracerTest />, { tracer });
    const ast = parser.parse("123");
    expect(ast).toEqual("123");
})

test("renderer.context", () => {

    function ContextTest() {
        return (
            <pattern action={({ digit, globalContext }) => {
                return globalContext.transform(digit);
            }}>
                <Digit label="digit" />
            </pattern>
        );
    }

    const parser = ReactPeg.render(<ContextTest />);
    const ast = parser.parse("1", {
        transform: (digit: string) => `digit value is ${digit}`
    });
    expect(ast).toEqual("digit value is 1");
})