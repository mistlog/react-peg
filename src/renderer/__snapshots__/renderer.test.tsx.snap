// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer.parse 1`] = `"Expected Digit but \\"a\\" found."`;

exports[`renderer: tracer 1`] = `
Object {
  "location": Object {
    "end": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "rule": "TracerTest",
  "type": "rule.enter",
}
`;

exports[`renderer: tracer 2`] = `
Object {
  "location": Object {
    "end": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "rule": "Digit",
  "type": "rule.enter",
}
`;

exports[`renderer: tracer 3`] = `
Object {
  "location": Object {
    "end": Object {
      "column": 2,
      "line": 1,
      "offset": 1,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "result": "1",
  "rule": "Digit",
  "type": "rule.match",
}
`;

exports[`renderer: tracer 4`] = `
Object {
  "location": Object {
    "end": Object {
      "column": 2,
      "line": 1,
      "offset": 1,
    },
    "start": Object {
      "column": 2,
      "line": 1,
      "offset": 1,
    },
  },
  "rule": "Digit",
  "type": "rule.enter",
}
`;

exports[`renderer: tracer 5`] = `
Object {
  "location": Object {
    "end": Object {
      "column": 3,
      "line": 1,
      "offset": 2,
    },
    "start": Object {
      "column": 2,
      "line": 1,
      "offset": 1,
    },
  },
  "result": "2",
  "rule": "Digit",
  "type": "rule.match",
}
`;

exports[`renderer: tracer 6`] = `
Object {
  "location": Object {
    "end": Object {
      "column": 3,
      "line": 1,
      "offset": 2,
    },
    "start": Object {
      "column": 3,
      "line": 1,
      "offset": 2,
    },
  },
  "rule": "Digit",
  "type": "rule.enter",
}
`;

exports[`renderer: tracer 7`] = `
Object {
  "location": Object {
    "end": Object {
      "column": 4,
      "line": 1,
      "offset": 3,
    },
    "start": Object {
      "column": 3,
      "line": 1,
      "offset": 2,
    },
  },
  "result": "3",
  "rule": "Digit",
  "type": "rule.match",
}
`;

exports[`renderer: tracer 8`] = `
Object {
  "location": Object {
    "end": Object {
      "column": 4,
      "line": 1,
      "offset": 3,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "result": "123",
  "rule": "TracerTest",
  "type": "rule.match",
}
`;
